[DOCS] Adding GenAI Use Cases (#27062)
openvinotoolkit/openvino: docs/articles_en/learn-openvino/llm_inference_guide/genai-guide/genai-use-cases.rst

GenAI Use Cases
=====================

> This article provides several use case scenarios for the inference on
> Generative AI Models. The applications presented in the code samples below

Reviewer: ```suggestion
This article provides several use case scenarios for Generative AI model
inference. The applications presented in the code samples below
```

only require minimal configuration, like setting an inference device. Feel free
to explore and modify the source code as you need.

> Using GenAI in Text to Image Conversion

Reviewer: ```suggestion
Using GenAI for Text-to-Image generation
```

> #######################################

Reviewer: ```suggestion
################################################
```

Examples below demonstrate inference on text-to-image models, like Stable Diffusion
1.5, 2.1, and LCM, with a text prompt as input. The :ref:`main.cpp <maincpp>`
sample shows basic usage of the ``Text2ImagePipeline`` pipeline.
:ref:`lora.cpp <loracpp>` shows how to apply LoRA adapters to the pipeline.

.. tab-set::

   .. tab-item:: C++
      :sync: cpp

      .. tab-set::

         .. tab-item:: main.cpp
            :name: maincpp

            .. code-block:: cpp

>                int32_t main(int32_t argc, char* argv[]) try {

Reviewer 2: If you provide `main()` you probably should add `#include`s. Knowing what to include is more important when knowing how to wrap the code into `main()`. Python imports are added below after all.

                   OPENVINO_ASSERT(argc == 3, "Usage: ", argv[0], " <MODEL_DIR> '<PROMPT>'");

                   const std::string models_path = argv[1], prompt = argv[2];
                   const std::string device = "CPU";  // GPU, NPU can be used as well.

                   ov::genai::Text2ImagePipeline pipe(models_path, device);
                   ov::Tensor image = pipe.generate(prompt,
                       ov::genai::width(512),
                       ov::genai::height(512),
                       ov::genai::num_inference_steps(20),
                       ov::genai::num_images_per_prompt(1));

                   // Saves images with a `num_images_per_prompt` name pattern.
                   imwrite("image_%d.bmp", image, true);

                   return EXIT_SUCCESS;
               } catch (const std::exception& error) {
                   try {
                       std::cerr << error.what() << '\n';
                   } catch (const std::ios_base::failure&) {}
                   return EXIT_FAILURE;

>                } catch (...) {

Reviewer 2: I guess this is a nice opportunity to remove try catch which is normally required by SDL.

Author: I have updated the code snippets:

https://github.com/openvinotoolkit/openvino/pull/27062/commits/08e4a066c128c5e0ce4daca9d6e14488132070ba

may prove beneficial. The ``start_chat`` and ``finish_chat`` chat-specific methods are used to
mark a conversation session, as shown in the samples below:

.. tab-set::

   .. tab-item:: Python
      :sync: py

      .. code-block:: python

         import openvino_genai as ov_genai
         pipe = ov_genai.LLMPipeline(model_path)

         pipe.set_generation_config({'max_new_tokens': 100)

         pipe.start_chat()
         while True:
            print('question:')
            prompt = input()
            if prompt == 'Stop!':
               break
            print(pipe.generate(prompt))
         pipe.finish_chat()

      For more information, refer to the
      `Python sample <https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/python/chat_sample/README.md>`__.

   .. tab-item:: C++
      :sync: cpp

      .. code-block:: cpp

         int main(int argc, char* argv[]) {
            std::string prompt;

            std::string model_path = argv[1];
            ov::genai::LLMPipeline pipe(model_path, "CPU");

            ov::genai::GenerationConfig config = pipe.get_generation_config();
            config.max_new_tokens = 100;

>             pipe.set_generation_config(config)

Reviewer 3: is there a ; missing?

Author: I suppose it is. The code moved from the main article seems to be outdated and I believe it should be changed to https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/cpp/chat_sample/chat_sample.cpp#L6C1-L33C2

@andrei-kochin can you confirm?

Reviewer 4: @sgolebiewski-intel yes, let's align with the sample

Author: I have aligned it:
https://github.com/openvinotoolkit/openvino/pull/27062/commits/08e4a066c128c5e0ce4daca9d6e14488132070ba

                   std::cout << "Generating image with LoRA adapters applied, resulting image will be in lora.bmp\n";
                   ov::Tensor image = pipe.generate(prompt,
                       ov::genai::random_generator(std::make_shared<ov::genai::CppStdGenerator>(42)),
                       ov::genai::width(512),
                       ov::genai::height(896),
                       ov::genai::num_inference_steps(20));
                   imwrite("lora.bmp", image, true);

                   std::cout << "Generating image without LoRA adapters applied, resulting image will be in baseline.bmp\n";
                   image = pipe.generate(prompt,
                       ov::genai::adapters(),  // Passing adapters as generation overrides set in the constructor; adapters() means no adapters.
                       ov::genai::random_generator(std::make_shared<ov::genai::CppStdGenerator>(42)),
                       ov::genai::width(512),
                       ov::genai::height(896),
                       ov::genai::num_inference_steps(20));
                   imwrite("baseline.bmp", image, true);

                   return EXIT_SUCCESS;

>                }

Reviewer 5: Can we prepare less bulky code snippet to be used as on-line illustration? Is there a goal to use exactly the same code as in samples? I think we can make reader life easier with a more distilled code.

      For more information, refer to the
      `C++ sample <https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/cpp/text2image/README.md>`__

Using GenAI in Speech Recognition
#################################

The application, shown in code samples below, performs inference on speech
recognition Whisper Models. The samples include the ``WhisperPipeline`` class
and use audio files in WAV format at a sampling rate of 16 kHz as input.

.. tab-set::

   .. tab-item:: Python
      :sync: cpp

      .. code-block:: python

>          import argparse

Reviewer 6: I would suggest using a more universal code here and in other examples that can be used both in python and from the command line, for example:

``` python
import argparse
import openvino_genai
import librosa

def read_wav(filepath):
    raw_speech, samplerate = librosa.load(filepath, sr=16000)
    return raw_speech.tolist()

def infer(model_dir: str, wav_file_path: str):
    raw_speech = read_wav(wav_file_path)
    pipe = openvino_genai.WhisperPipeline(model_dir)

    def streamer(word: str) -> bool:
        print(word, end="")
        return False

    result = pipe.generate(
        raw_speech,
        max_new_tokens=100,
        # The 'task' and 'language' parameters are supported for multilingual models only.
        language="<|en|>",
        task="transcribe",
        return_timestamps=True,
        streamer=streamer,
    )

    print()
    for chunk in result.chunks:
        print(f"timestamps: [{chunk.start_ts}, {chunk.end_ts}] text: {chunk.text}")
    

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("model_dir")
    parser.add_argument("wav_file_path")
    args = parser.parse_args()
    
    infer(args.model_dir, args.wav_file_path)

if __name__ == '__main__':
    main()
```
 or just 

```python 
import openvino_genai
import librosa

def read_wav(filepath):
    raw_speech, samplerate = librosa.load(filepath, sr=16000)
    return raw_speech.tolist()

def infer(model_dir: str, wav_file_path: str):
    raw_speech = read_wav(wav_file_path)
    pipe = openvino_genai.WhisperPipeline(model_dir)

    def streamer(word: str) -> bool:
        print(word, end="")
        return False

    result = pipe.generate(
        raw_speech,
        max_new_tokens=100,
        # The 'task' and 'language' parameters are supported for multilingual models only.
        language="<|en|>",
        task="transcribe",
        return_timestamps=True,
        streamer=streamer,
    )

    print()
    for chunk in result.chunks:
        print(f"timestamps: [{chunk.start_ts}, {chunk.end_ts}] text: {chunk.text}")
```

This will help you reuse the example code in different scenarios without having to rewrite it.

Author: @andrei-kochin, @Wovchena, @slyalin How about the suggestions above? Do we include them to the documentation? If so, which is preferrable?

Reviewer 4: @sgolebiewski-intel  I vote for the option 2 (without the main)

             parser = argparse.ArgumentParser()
             parser.add_argument("model_dir")
             parser.add_argument("wav_file_path")
             args = parser.parse_args()

             raw_speech = read_wav(args.wav_file_path)

             pipe = openvino_genai.WhisperPipeline(args.model_dir)

             def streamer(word: str) -> bool:
                 print(word, end="")
                 return False

             result = pipe.generate(
                 raw_speech,
                 max_new_tokens=100,
                 # The 'task' and 'language' parameters are supported for multilingual models only.
                 language="<|en|>",
                 task="transcribe",
                 return_timestamps=True,
                 streamer=streamer,
             )

             print()

             for chunk in result.chunks:
                 print(f"timestamps: [{chunk.start_ts}, {chunk.end_ts}] text: {chunk.text}")

      For more information, refer to the
      `Python sample <https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/python/whisper_speech_recognition/README.md>`__.

   .. tab-item:: C++
      :sync: cpp

      .. code-block:: cpp

>          int main(int argc, char* argv[]) {

Reviewer 4: Please use the built-in code from the code sample

Author: Can we reach a decision here? @slyalin I am taking into consideration your comment regarding distilling the code snippets.

Author: Currently, we see 2 possible solutions.

1. We can keep the whole built-in code, while commenting out some lines of code, thus leaving the part on which users can focus.
![image](https://github.com/user-attachments/assets/3af12a8c-3e30-4f11-a3d0-0a1eb4ac7608)

2. We keep the distilled code and provide a link to the source example. 

![image](https://github.com/user-attachments/assets/1e56b505-9173-49b4-b2d8-fa3cc7a13b86)

Please, share your thoughts.

Reviewer 2: Or is it called `built-in code`?

Author: > Is there an option to embed the sample from a repo? Avoiding copy would be the best solution

As far as I know, it is impossible to directly embed files outside the scope of `openvino/docs` directory. We can make reference to such files.

Reviewer 2: In that case having such commented out code is weird to me, so distilled code looks is better. @andrei-kochin will take a look once more.

Would there be some mechanism ensuring the code snippets are kept up to date?

Author: > In that case having such commented out code is weird to me, so distilled code looks is better. @andrei-kochin will take a look once more.
> 
> Would there be some mechanism ensuring the code snippets are kept up to date?

For now, I believe, these snippets will be listed under usual maintenance items to keep them up to date.

Reviewer 4: @sgolebiewski-intel I agree to keep only inline comment related to the device. Others are not required and can be removed
